### PR TITLE
Allow typing in date widget

### LINKED ISSUE #1178: Bug: Date Widgets does not allow typing in a date.
**- Do you want to request a *feature* or report a *bug*?**
Bug

**- What is the current behavior?**
Date and DateTime widgets updates (?) after every character you type, often overwriting what your trying to enter.

**- If the current behavior is a bug, please provide the steps to reproduce.**
For example, let's say I have 01/06/2018 in the widget. I want to change it to 2008. However, Placing the cursor at the end and hitting backspace immediately updates the widget so it says 01/06/0201. Also hitting backspace again updates it to 01/06/2020 which adds confusion (where did the 2 at the start come from? shouldn't it say 0020 to be consistent?).

Second example, let's say I want to type in 01/05/2004. I delete the contents of the widget and type 0. It immediately fills in with 01/01/2000 completely unbidden and prevents any more typing at the end. I have to move the cursor to each character and try to delete then type in my number. 

**- What is the expected behavior?**
While trying to type in a date, the widget should only update the content and reformat once you click out of the field or hit enter/tab.

**- Please mention your CMS, node.js, and operating system version.**

**- Please link or paste your `config.yml` below if applicable.**
Just a standard date field definition.
```
      - {required: false,  label: "Birth Date", name: "birth_date", widget: "date"}
```

**- Notes.**

This is a fairly frustrating bug. It does have some rhyme and reason, but on the surface looks like it just changes your input randomly. Currently you can set the date by copy and pasting in a valid date format (like 01/05/2004) or using the graphical picker. Both these work fine in my experience. But many will want to just type in dates directly from memory which invokes this bug and makes them frustrated.

Apologies if this is a known issue. I searched and didn't find anything.

Thank you.

### PR DESCRIPTION
**Pull request by Dammmien**
Fixes #1178 : Allow typing in date widget

**- Summary**

Allow typing in date widget by setting the date only if the format is valid

**- Test plan**

Test date widget behavior is ok

**- Description for the changelog**

Allow typing in date widget

### CODE REVIEW

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit 67be56c1cd47d26163f196e3e0b877ad9574dbea

https://deploy-preview-1247--cms-demo.netlify.com

**Comment by verythorough** [comment]
Deploy preview for *netlify-cms-www* ready!

Built with commit 67be56c1cd47d26163f196e3e0b877ad9574dbea

https://deploy-preview-1247--netlify-cms-www.netlify.com

File: src/components/EditorWidgets/Date/DateControl.js [hunk 1]
```diff
@@ -36,9 +36,8 @@ export default class DateControl extends React.Component {
 
   handleChange = datetime => {
     const { onChange } = this.props;
-    if (!this.format || datetime === '') {
```
Comment (tech4him1) [inline src/components/EditorWidgets/Date/DateControl.js]: We do still need to handle the empty case here, in case the user wants to empty the field.

**Comment by tech4him1** [comment]
This looks great -- I think reverting to the previous value if the date is invalid is reasonable. Do you think we need to warn the user at that point, or just revert it? For simple changes, the user wouldn't necessarily notice that it was reverted (e.g. entered `SM` instead of `AM`).

**Comment by Dammmien** [comment]
@tech4him1 The empty case is now supported and the user is warned onBlur that the date is invalid.

**Comment by erquhart** [comment]
This is definitely an improvement, thanks @Dammmien!

I do find the warning jarring, especially considering the case where a user types in a date, and it's almost guaranteed to be the wrong format, then they get a generic warning and the field goes blank.

I'd like us to consider attempting to `moment.parse` whatever they type, and use the result if it's valid. Moment is really good at parsing dates, and while much of the fast and loose parsing is deprecated, their 3.0 milestone is years old and looks unlikely to ever come to fruition. In the future we could also move to date-fns which seems quite happy to parse strings as well.

Thoughts?

**Comment by tech4him1** [comment]
Also, if we do choose to use `moment.parse`, I don't see any problem with just using the functions in `react-datetime` directly to parse and check validity. We should simply be able to set the `dateFormat` on it.

**Comment by Dammmien** [comment]
@erquhart , @tech4him1
I pushed a new version, where the input is parsed with `moment(...)` if the date is valid it's used otherwise it's reverted to the previous value.
What do you think about it ?

File: src/components/EditorWidgets/Date/DateControl.js [hunk 2]
```diff
@@ -34,16 +34,35 @@ export default class DateControl extends React.Component {
     }
   }
 
+  // Date is valid if datetime is a moment or Date object otherwise it's a string.
+  // Handle the empty case, if the user wants to empty the field.
+  isValid = datetime => (moment.isMoment(datetime) || datetime instanceof Date || datetime === '');
```
Comment (tech4him1) [inline src/components/EditorWidgets/Date/DateControl.js]: `isValid` has a specific use for widget controls, which we're not trying to use here, so can we change this to something else, maybe `isValidDate`?

**Comment by tech4him1** [comment]
@erquhart Just so I understand -- we're not warning users if the date is invalid because of why? Now that we are parsing with Moment it should rarely be invalid.

**Comment by erquhart** [comment]
Right, so you'd have to type something pretty odd to get a blank field.

**Comment by tech4him1** [comment]
Is there any reason we can't throw the warning only if the value is invalid, since that should be rare now?

**Comment by erquhart** [comment]
It is clearer that way - your call, I'm good with either.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Dammmien** [comment]
@tech4him1 I changed `isValid` to `isValidDate` and restored the warning when the value can't be parsed by Moment.

**Pull request merged**